lib/test.js

```diff
@@ -53,6 +53,77 @@ function plan(count) {
 	this.plan(count, captureStack(this.plan));
 }
 
+function tryTest(fn) {
+	const args = [];
+	for (let i = 1, len = arguments.length; i < len; i++) {
+		args.push(arguments[i]);
+	}
+
+	let commitCalled = false;
+	let discarded = false;
+	let finished = false;
+
+	const commitFn = (ret, toCommit) => {
+		if (commitCalled) {
+			if (discarded) {
+				this.saveFirstError(new Error('The discard() was already called'));
+			} else {
+				this.saveFirstError(new Error('The commit() was already called'));
+			}
+			return;
+		}
+		commitCalled = true;
+
+		if (ret.passed) {
+			this.countFinishedAttemptAssertion(toCommit);
+		} else {
+			this.countFailedAttemptAssertion(ret.error, toCommit);
+		}
+
+		if (toCommit) {
+			for (const log of ret.logs) {
+				this.addLog(log);
+			}
+		}
+	};
+
+	const running = this
+		.forAttempt(t => fn.apply(null, [t].concat(args)))
+		.run();
+
+	this.addPendingAttemptAssertion();
+
+	const finishedRunning = running.then(ret => {
+		if (discarded) {
+			return null;
+		}
+
+		finished = true;
+		return Object.assign({}, ret, {
+			commit: () => commitFn(ret, true),
+			discard: () => {
+				discarded = true;
+				return commitFn(ret, false);
+			}
+		});
+	});
+
+	finishedRunning.discard = () => {
+		if (finished) {
+			this.saveFirstError(new Error('Attempt is already resolved'));
```

Comment:
I think that's fine, it's only a problem if it was committed rather than discarded.

Reply:
So should it not throw an error and simply return?

Reply:
No, I meant that it shouldn't matter whether you discard the promise or the result, even if the promise has been fulfilled.